### DISCUSSION TITLE
Introducing a Global Variable for Author/Reader Mode

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6074] [Ideas]
There was time to time discussion on forum to be able to hide some tools from UI when a TW is put on the server to public access or when distributed to the end users.

I have used a mechanism in [Utility plugin](https://kookma.github.io/TW-Utility/#demo%2Futility-author-tools) and have seen in other plugin/tools like the Eric Shulman [Inside Tiddlywiki book](https://tiddlytools.com/InsideTW/)!

Suggestion:

- add a global variable like `$:/config/admin` variable with value `yes`, `no`
- a UI lets decide which elements are admin tool and shall be hidden in reader mode
- allow plugin developer to use this variable to hide admin tools in their plugin 
- allow content creators to use the variable to hide some admin tools like (edit button/delete button) etc.

**Comment by Jermolene** [comment]
Hi @kookma we already have support for this, but it is only implemented in the client-server configuration:

* the core defines a status tiddler `$:/status/IsReadOnly` that reflects whether the user is able to make changes or not
* The tiddlywebadaptor includes [a stylesheet](https://github.com/Jermolene/TiddlyWiki5/blob/master/plugins/tiddlywiki/tiddlyweb/readonly-styles.tid) that hides editing controls when in read only mode

The reason the core doesn't do something similar for the single file configuration is there is no general way for the core to tell whether the wiki should be treated as read-only.

**Reply by kookma** [reply]
> The reason the core doesn't do something similar for the single file configuration is there is no general way for the core to tell whether the wiki should be treated as read-only.

does it make sense to manually set `$:/status/IsReadOnly` to `yes` or `no` ? doing this through a shortcut will allow us to distribute a wiki to end user with much less complexity!

**Reply by saqimtiaz** [reply]
@kookma that is essentially what the steps I described below are doing.

**Comment by saqimtiaz** [comment]
I set up a wiki for a colleague with a keyboard shortcut to switch between a read-only viewer mode that hides most of the complexity of the interface, and an editor view that is the default TiddlyWiki view.

The steps as I recall them:

- a global keyboard shortcut that toggles the value of `$:/status/IsReadOnly` between `yes` and `no`, where the default is to change to `no`, and also switches page layouts
- an alternative page layout that clones the core page template but then removes the sidebar and replaces it with a simpler one and makes some other layout changes.
- a `publishFilter` that excludes `$:/status/IsReadOnly `from being saved.
- the CSS from `$:/plugins/tiddlywiki/tiddlyweb/readonly` and some custom CSS.

With this in place the wiki always loads as readOnly, can easily be switched to editing mode, and will never accidentally be saved in a manner that it opens in editing mode. Hope this helps.

**Reply by kookma** [reply]
> * a global keyboard shortcut that toggles the value of `$:/status/IsReadOnly` between `yes` and `no`, where the default is to change to `no`, and also switches page layouts

This is interesting! my concern is to have a standard core tiddler to be used by plugin developers! So, for example I can use it to hide some UI in Commander plugin, and so on



> * a `publishFilter` that excludes `$:/status/IsReadOnly `from being saved.

Is this a node.js wiki or a single file?

By the way I like your approach! and I will think to improve the Utility plugin if I cannot use `$:/status/IsReadOnly` in single wiki file as Jeremy explained!

**Reply by saqimtiaz** [reply]
> Is this a node.js wiki or a single file?

Single file

> if I cannot use `$:/status/IsReadOnly` in single wiki file as Jeremy explained!

You can use it, that is what my description above uses as well, you just need to set the value yourself as the core does not know whether the wiki should be read only or not and cannot create the tiddler for you when the wiki loads.

**Reply by AnthonyMuscio** [reply]
> * a `publishFilter` that excludes `$:/status/IsReadOnly `from being saved.

[See my other contribution](https://talk.tiddlywiki.org/t/introducing-a-global-variable-for-author-reader-mode/902/2?u=tw_tones) 

I have used a system that does this for a serial editing process (I have 90% completed) that does this. for example have a tiddler called $:/config/local-only containing `$:/status/IsReadOnly` and other tiddlers.

I currently use local storage, but cookies can be used 
- So the local storage filter ends with
`[subfilter{$:/config/local-only}] ` this ensures $:/status/IsReadOnly is only saved in local storage
- But then its complement set the publish filter
 `\define publishFilter() -[subfilter{$:/PSaT/CIO/local-only-tiddlers}]` is not saved in the wiki
- However I also toggle tiddlers for autosave and the filter that sets the dirty state. eg ` -[subfilter{$:/PSaT/CIO/local-only-tiddlers}]`

Then to support this I ensure all the tiddlers that define the read only state are set in a plugin so they are shadow tiddlers. Thus the default will always be read only unless the settings are switched to edit, and this switching occurs only in local storage. 

That is the edit mode is not committed to disk even when other tiddlers are saved during edit. Otherwise a new visitor could open the wiki in edit mode while I am editing it.

**Comment by AnthonyMuscio** [comment]
I have responded to this in [talk.tiddlywiki.org]([url](https://talk.tiddlywiki.org/t/introducing-a-global-variable-for-author-reader-mode/902/3?u=tw_tones)) based of extensive experience with this subject.

In relation to $:/status/IsReadOnly it would just be an other value associated with say a read-only mode.
